PCH: Enhance Request Handling in Provider Classes

vaurdan opened the pull request:
## Description
This PR refactors all provider classes to inherit from a newly introduced `BaseProvider`. This new base class equips each provider with an enhanced method for fetching data from the API that includes support for cancelling ongoing requests using the `AbortController`.

This update transitions provider methods from static to instance methods, which allows each provider to maintain its own map of `AbortController` instances, one per request. This architecture ensures that cancelling one request does not inadvertently cancel others. While not currently implemented, this setup facilitates future enhancements, such as assigning unique abort IDs for selectively cancelling requests in complex operations like smart linking across multiple paragraphs.

This architecture not only simplifies the management of request cancellations but also positions the codebase for more sophisticated request handling scenarios in the future.

## Motivation and Context
- Implements cancellable requests, thereby resolving issue #1928.
- Enhances the reliability of the codebase and reduces redundancy by utilizing inheritance for common provider functionalities.

## How has this been tested?
Each provider was tested individually through manual checks, end-to-end tests, and unit tests with Jest.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a base class for improved data fetching with request cancellation capabilities.
	- Enhanced smart linking and title suggestion functionalities by utilizing a singleton pattern for better resource management.

- **Bug Fixes**
	- Removed the `isDismissible` property from certain UI components to streamline user notifications.

- **Documentation**
	- Standardized JSDoc comments across various components for better consistency and readability.

- **Refactor**
	- Updated several components to use singleton access methods, aligning with modern best practices.

- **Chores**
	- Updated dependencies to improve stability and performance of testing frameworks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

vaurdan committed:
Create BaseProvider class and apply it to SmartLinkingProvider

vaurdan committed:
Apply new BaseProvider to PerformanceStats provider

vaurdan committed:
Apply new BaseProvider to RelatedPostsProvider

vaurdan committed:
Apply new BaseProvider to TitleSuggestionsProvider

vaurdan committed:
Apply new BaseProvider to DashboardWidgetProvider

vaurdan committed:
Refactor BaseProvider to use non-static methods instead.

vaurdan committed:
Clean-up abort controller after each request.

vaurdan committed:
Remove testing code

vaurdan committed:
Merge branch 'refs/heads/develop' into update/ch-provider-refactor

# Conflicts:
#	build/content-helper/dashboard-widget.asset.php
#	build/content-helper/editor-sidebar.asset.php

vaurdan committed:
Rebuild assets

vaurdan commented:
@coderabbitai review

acicovic committed:
Adjust some whitespace/comments

acicovic committed:
Adjust some whitespace/comments

acicovic commented on src/content-helper/common/base-provider.tsx:
```diff
@@ -0,0 +1,174 @@
+	data: T;
+}
+
+type GetAbortControllerResult = {
```
We can add a comment here if we think it would be useful.

acicovic reviewed (approved):
Thank you for your work and for the great PR description!

Approving with a minor comment.

acicovic committed:
TypeScript DocBlocks: Remove spaces in braces

acicovic committed:
Merge branch 'develop' into update/ch-provider-refactor

dependabot[bot] committed:
build(deps-dev): bump @testing-library/react from 15.0.2 to 15.0.4

Bumps [@testing-library/react](https://github.com/testing-library/react-testing-library) from 15.0.2 to 15.0.4.
- [Release notes](https://github.com/testing-library/react-testing-library/releases)
- [Changelog](https://github.com/testing-library/react-testing-library/blob/main/CHANGELOG.md)
- [Commits](https://github.com/testing-library/react-testing-library/compare/v15.0.2...v15.0.4)

---
updated-dependencies:
- dependency-name: "@testing-library/react"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

acicovic committed:
PCH Editor Sidebar: Update isDismissible in Notices

acicovic committed:
Remove unnecessary empty line

vaurdan committed:
Add JSDoc comment to `GetAbortControllerResult`

vaurdan committed:
Merge branch 'develop' into update/ch-provider-refactor

acicovic commented on src/content-helper/editor-sidebar/title-suggestions/provider.ts:
```diff
@@ -1,75 +1,75 @@
+	 * @param { ToneProp }    tone    The tone to use when generating the titles.
+	 * @param { PersonaProp } persona The persona to use when generating the titles.
 	 *
+	                                  <<<<<<< HEAD
+	                                  <<<<<<< HEAD
+	 * @return { Promise<string[]> } The resulting list of titles.
+	  =======
 	 * @return {Promise<string[]>} The resulting list of titles.
+	  >>>>>>> 3696940e (TypeScript DocBlocks: Remove spaces in braces)
+	  =======
+	 * @return {Promise<string[]>} The resulting list of titles.
+	  >>>>>>> develop
```
Some funky stuff happening here 🙂

acicovic replied:
It also seems some additional things were reverted in https://github.com/Parsely/wp-parsely/pull/2401/commits/9396e93e48a58cb7def1b091d9638ddb15faf1bd? A bit strange.

acicovic replied:
The revert seems to be mostly the changes of https://github.com/Parsely/wp-parsely/pull/2401/commits/e9d8c5f81ad6fa1af658792d4c305bbfc28241d2.

vaurdan replied:
This is weird! Let me clean this up.

vaurdan committed:
Clean-up docblock merge mess

vaurdan commented:
@acicovic the merge mess should be fixed! Not sure what happened there :D 

I took a look at the change set, and it looks alright. But just in case, feel free to take another look and let me know if I have 🟢 to merge!

vaurdan committed:
Remove extra spaces.

acicovic reviewed (approved):
I think it's OK now.
